syntax = "proto3";

package trainer;

import "google/protobuf/empty.proto";

message TrainRequest {
  reserved 2;
  reserved "dataset_uri";

  string              model_name  = 1;
  repeated string     knowledge   = 6;
  map<string,string>  hyperparams = 3;
  uint32              gpu_id      = 4;
  string              job_id      = 5;

  string              name        = 7;
  string              description = 8;
  string              instruction = 9;
}
/* ───────── Streaming progress update ───────── */
message TrainProgress {
  enum Stage {
    RECEIVED  = 0;
    PREPARING = 1;
    READY     = 2;
    TRAINING  = 3;
    COMPLETE  = 4;
    FAILED    = 5;
  }

  uint32 step        = 1;
  uint32 total_steps = 2;
  double loss        = 3;
  double metric      = 4;

  Stage  stage       = 5;
  string job_id      = 6;
  string message     = 7;
}

/* ───────── Final reply (after stream ends) ───────── */
message TrainReply {
  bool   success      = 1;
  string message      = 2;
  string artifact_uri = 3;
}

/* ───────── Services ───────── */
service Trainer {
  rpc StartJob (TrainRequest) returns (stream TrainProgress);
  rpc CancelJob (TrainRequest) returns (google.protobuf.Empty);
}
